In [2]:
#import dependencies for Data Exploration
import pandas as pd
import numpy as np


In [3]:
#Set display option to view entire dataset
pd.set_option('display.max_columns', 200)
pd.set_option('display.width', 300)
pd.set_option('display.max_rows', 160)

In [4]:
# Read raw dataset from csv file
data = pd.read_csv("accepted_2007_to_2018Q4.csv", low_memory=False)
data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade sub_grade                    emp_title emp_length home_ownership  annual_inc verification_status   issue_d loan_status pymnt_plan                                                url desc  \
0  68407277        NaN     3600.0       3600.0           3600.0   36 months     13.99       123.03     C        C4                      leadman  10+ years       MORTGAGE     55000.0        Not Verified  Dec-2015  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
1  68355089        NaN    24700.0      24700.0          24700.0   36 months     11.99       820.28     C        C1                     Engineer  10+ years       MORTGAGE     65000.0        Not Verified  Dec-2015  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
2  68341763        NaN    20000.0      20000.0          20000.0   60 months     10.78       432.66     B        B4                 truck driver  10+ years       MORTGAGE     63000.0        Not Verified  Dec-2015  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
3  66310712        NaN    35000.0      35000.0          35000.0   60 months     14.85       829.90     C        C5  Information Systems Officer  10+ years       MORTGAGE    110000.0     Source Verified  Dec-2015     Current          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
4  68476807        NaN    10400.0      10400.0          10400.0   60 months     22.45       289.91     F        F1          Contract Specialist    3 years       MORTGAGE    104433.0     Source Verified  Dec-2015  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   

              purpose               title zip_code addr_state    dti  delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0  debt_consolidation  Debt consolidation    190xx         PA   5.91          0.0         Aug-2003           675.0            679.0             1.0                    30.0                     NaN       7.0      0.0     2765.0        29.7       13.0                   w       0.00           0.00   
1      small_business            Business    577xx         SD  16.06          1.0         Dec-1999           715.0            719.0             4.0                     6.0                     NaN      22.0      0.0    21470.0        19.2       38.0                   w       0.00           0.00   
2    home_improvement                 NaN    605xx         IL  10.78          0.0         Aug-2000           695.0            699.0             0.0                     NaN                     NaN       6.0      0.0     7869.0        56.2       18.0                   w       0.00           0.00   
3  debt_consolidation  Debt consolidation    076xx         NJ  17.06          0.0         Sep-2008           785.0            789.0             0.0                     NaN                     NaN      13.0      0.0     7802.0        11.6       17.0                   w   15897.65       15897.65   
4      major_purchase      Major purchase    174xx         PA  25.37          1.0         Jun-1998           695.0            699.0             3.0                    12.0                     NaN      12.0      0.0    21929.0        64.5       35.0                   w       0.00           0.00   

    total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pull_d  last_fico_range_high  last_fico_range_low  collections_12_mths_ex_med  mths_since_last_major_derog  policy_code  \
0   4421.723917          4421.72          3600.00         821.72                 0.0         0.0                      0.0     Jan-2019           122.67    

In [5]:
# Description of dataset
data.describe()

member_id     loan_amnt   funded_amnt  funded_amnt_inv      int_rate   installment    annual_inc           dti   delinq_2yrs  fico_range_low  fico_range_high  inq_last_6mths  mths_since_last_delinq  mths_since_last_record      open_acc       pub_rec     revol_bal    revol_util  \
count        0.0  2.260668e+06  2.260668e+06     2.260668e+06  2.260668e+06  2.260668e+06  2.260664e+06  2.258957e+06  2.260639e+06    2.260668e+06     2.260668e+06    2.260638e+06            1.102166e+06           359156.000000  2.260639e+06  2.260639e+06  2.260668e+06  2.258866e+06   
mean         NaN  1.504693e+04  1.504166e+04     1.502344e+04  1.309283e+01  4.458068e+02  7.799243e+04  1.882420e+01  3.068792e-01    6.985882e+02     7.025884e+02    5.768354e-01            3.454092e+01               72.312842  1.161240e+01  1.975278e-01  1.665846e+04  5.033770e+01   
std          NaN  9.190245e+03  9.188413e+03     9.192332e+03  4.832138e+00  2.671735e+02  1.126962e+05  1.418333e+01  8.672303e-01    3.301038e+01     3.301124e+01    8.859632e-01            2.190047e+01               26.464094  5.640861e+00  5.705150e-01  2.294831e+04  2.471307e+01   
min          NaN  5.000000e+02  5.000000e+02     0.000000e+00  5.310000e+00  4.930000e+00  0.000000e+00 -1.000000e+00  0.000000e+00    6.100000e+02     6.140000e+02    0.000000e+00            0.000000e+00                0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%          NaN  8.000000e+03  8.000000e+03     8.000000e+03  9.490000e+00  2.516500e+02  4.600000e+04  1.189000e+01  0.000000e+00    6.750000e+02     6.790000e+02    0.000000e+00            1.600000e+01               55.000000  8.000000e+00  0.000000e+00  5.950000e+03  3.150000e+01   
50%          NaN  1.290000e+04  1.287500e+04     1.280000e+04  1.262000e+01  3.779900e+02  6.500000e+04  1.784000e+01  0.000000e+00    6.900000e+02     6.940000e+02    0.000000e+00            3.100000e+01               74.000000  1.100000e+01  0.000000e+00  1.132400e+04  5.030000e+01   
75%          NaN  2.000000e+04  2.000000e+04     2.000000e+04  1.599000e+01  5.933200e+02  9.300000e+04  2.449000e+01  0.000000e+00    7.150000e+02     7.190000e+02    1.000000e+00            5.000000e+01               92.000000  1.400000e+01  0.000000e+00  2.024600e+04  6.940000e+01   
max          NaN  4.000000e+04  4.000000e+04     4.000000e+04  3.099000e+01  1.719830e+03  1.100000e+08  9.990000e+02  5.800000e+01    8.450000e+02     8.500000e+02    3.300000e+01            2.260000e+02              129.000000  1.010000e+02  8.600000e+01  2.904836e+06  8.923000e+02   

          total_acc     out_prncp  out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee    recoveries  collection_recovery_fee  last_pymnt_amnt  last_fico_range_high  last_fico_range_low  collections_12_mths_ex_med  mths_since_last_major_derog  \
count  2.260639e+06  2.260668e+06   2.260668e+06  2.260668e+06     2.260668e+06     2.260668e+06   2.260668e+06        2.260668e+06  2.260668e+06             2.260668e+06     2.260668e+06          2.260668e+06         2.260668e+06                2.260523e+06                580775.000000   
mean   2.416255e+01  4.206891e+03   4.205965e+03  1.208256e+04     1.206439e+04     9.505772e+03   2.431388e+03        1.518453e+00  1.438791e+02             2.398257e+01     3.429346e+03          6.876610e+02         6.755397e+02                1.814580e-02                    44.164220   
std    1.198753e+01  7.343239e+03   7.342333e+03  9.901383e+03     9.896992e+03     8.321852e+03   2.679738e+03        1.184159e+01  7.481640e+02             1.312256e+02     6.018248e+03          7.297044e+01         1.110976e+02                1.508131e-01                    21.533121   
min    1.000000e+00  0.000000e+00   0.000000e+00  0.000000e+00     0.000000e+00     0.000000e+00   0.000000e+00       -9.500000e-09  0.000000e+00             0.000000e+00     0.000000e+00          0.000000e+00         0.000000e+00                0.00

In [6]:
# Description of dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260701 entries, 0 to 2260700
Columns: 151 entries, id to settlement_term
dtypes: float64(113), object(38)
memory usage: 2.5+ GB


In [7]:
#Creating a copy of raw data file
data_v1 = data.copy()

In [8]:
# Explore main categorical value column to determine which "loan_status" data to include in ML dataset
data_v1.loan_status.value_counts()

Fully Paid                                             1076751
Current                                                 878317
Charged Off                                             268559
Late (31-120 days)                                       21467
In Grace Period                                           8436
Late (16-30 days)                                         4349
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     40
Name: loan_status, dtype: int64

In [9]:
# Count sum of "na's" in each column 
data_v1.isna().sum()

id                                                  0
member_id                                     2260701
loan_amnt                                          33
funded_amnt                                        33
funded_amnt_inv                                    33
term                                               33
int_rate                                           33
installment                                        33
grade                                              33
sub_grade                                          33
emp_title                                      167002
emp_length                                     146940
home_ownership                                     33
annual_inc                                         37
verification_status                                33
issue_d                                            33
loan_status                                        33
pymnt_plan                                         33
url                         

In [10]:
missing_data = data_v1.isna().sum()
missing_data

id                                                  0
member_id                                     2260701
loan_amnt                                          33
funded_amnt                                        33
funded_amnt_inv                                    33
term                                               33
int_rate                                           33
installment                                        33
grade                                              33
sub_grade                                          33
emp_title                                      167002
emp_length                                     146940
home_ownership                                     33
annual_inc                                         37
verification_status                                33
issue_d                                            33
loan_status                                        33
pymnt_plan                                         33
url                         

In [11]:
# Count of columns missing over 100,000 rows of data
missing_data_over100k = missing_data[missing_data > 100000].count()
missing_data_over100k

63

In [12]:
# Count of columns missing over 50,000 rows of data
missing_data_over50k = missing_data[missing_data > 50000].count()
missing_data_over50k

94

In [13]:
# Explore data types
data_v1.dtypes

id                                             object
member_id                                     float64
loan_amnt                                     float64
funded_amnt                                   float64
funded_amnt_inv                               float64
term                                           object
int_rate                                      float64
installment                                   float64
grade                                          object
sub_grade                                      object
emp_title                                      object
emp_length                                     object
home_ownership                                 object
annual_inc                                    float64
verification_status                            object
issue_d                                        object
loan_status                                    object
pymnt_plan                                     object
url                         

In [14]:
# Correlation table
data_v1.corr()

member_id  loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  annual_inc       dti  delinq_2yrs  fico_range_low  fico_range_high  inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc   pub_rec  revol_bal  revol_util  \
member_id                                         NaN        NaN          NaN              NaN       NaN          NaN         NaN       NaN          NaN             NaN              NaN             NaN                     NaN                     NaN       NaN       NaN        NaN         NaN   
loan_amnt                                         NaN   1.000000     0.999755         0.999038  0.098082     0.945625    0.197246  0.043542    -0.009277        0.110585         0.110584       -0.027840               -0.010849                0.016293  0.182229 -0.058772   0.316715    0.099078   
funded_amnt                                       NaN   0.999755     1.000000         0.999341  0.098097     0.945967    0.197207  0.043745    -0.009173        0.110164         0.110163       -0.028132               -0.010851                0.016290  0.182392 -0.058636   0.316643    0.099132   
funded_amnt_inv                                   NaN   0.999038     0.999341         1.000000  0.098129     0.945118    0.197025  0.044276    -0.009029        0.109871         0.109870       -0.030063               -0.009963                0.023945  0.182601 -0.058288   0.316024    0.099087   
int_rate                                          NaN   0.098082     0.098097         0.098129  1.000000     0.123949   -0.050583  0.124491     0.058909       -0.415995        -0.415991        0.188399               -0.043481               -0.010175 -0.010472  0.053381  -0.028880    0.262670   
installment                                       NaN   0.945625     0.945967         0.945118  0.123949     1.000000    0.190270  0.044841     0.002053        0.053291         0.053291       -0.002391               -0.020234                0.000616  0.171445 -0.047035   0.302816    0.120951   
annual_inc                                        NaN   0.197246     0.197207         0.197025 -0.050583     0.190270    1.000000 -0.082619     0.026134        0.037009         0.037008        0.020001               -0.030504               -0.055449  0.094377 -0.002543   0.190760    0.028207   
dti                                               NaN   0.043542     0.043745         0.044276  0.124491     0.044841   -0.082619  1.000000    -0.012212       -0.027929        -0.027931       -0.012189                0.013869                0.060020  0.186124 -0.026029   0.102147    0.115225   
delinq_2yrs                                       NaN  -0.009277    -0.009173        -0.009029  0.058909     0.002053    0.026134 -0.012212     1.000000       -0.178913        -0.178910        0.024260               -0.552022               -0.076988  0.048956 -0.021810  -0.026779   -0.000200   
fico_range_low                                    NaN   0.110585     0.110164         0.109871 -0.415995     0.053291    0.037009 -0.027929    -0.178913        1.000000         1.000000       -0.093276                0.102196                0.247489  0.018407 -0.188583   0.010711   -0.477001   
fico_range_high                                   NaN   0.110584     0.110163         0.109870 -0.415991     0.053291    0.037008 -0.027931    -0.178910        1.000000         1.000000       -0.093277                0.102196                0.247489  0.018405 -0.188580   0.010709   -0.476999   
inq_last_6mths                                    NaN  -0.027840    -0.028132        -0.030063  0.188399    -0.002391    0.020001 -0.012189     0.024260       -0.093276        -0.093277        1.000000                0.011811               -0.089614  0.138064  0.064950  -0.010130   -0.078520   
mths_since_last_delinq                            NaN  -0.010849    -0.010851        -0.009963 -0.043481    -0.020234   -0.030504  0.013869    -0.552022        0.102196         0.102196        0.0

In [18]:
duplicate_col_names = data_v1.corr().columns.values

for col, row in (data_v1.corr().abs() > 0.95).iteritems():
    print(col, duplicate_col_names[row.values])

member_id []
loan_amnt ['loan_amnt' 'funded_amnt' 'funded_amnt_inv']
funded_amnt ['loan_amnt' 'funded_amnt' 'funded_amnt_inv']
funded_amnt_inv ['loan_amnt' 'funded_amnt' 'funded_amnt_inv']
int_rate ['int_rate']
installment ['installment']
annual_inc ['annual_inc']
dti ['dti']
delinq_2yrs ['delinq_2yrs']
fico_range_low ['fico_range_low' 'fico_range_high']
fico_range_high ['fico_range_low' 'fico_range_high']
inq_last_6mths ['inq_last_6mths']
mths_since_last_delinq ['mths_since_last_delinq']
mths_since_last_record ['mths_since_last_record']
open_acc ['open_acc' 'num_sats']
pub_rec ['pub_rec']
revol_bal ['revol_bal']
revol_util ['revol_util']
total_acc ['total_acc']
out_prncp ['out_prncp' 'out_prncp_inv']
out_prncp_inv ['out_prncp' 'out_prncp_inv']
total_pymnt ['total_pymnt' 'total_pymnt_inv' 'total_rec_prncp']
total_pymnt_inv ['total_pymnt' 'total_pymnt_inv' 'total_rec_prncp']
total_rec_prncp ['total_pymnt' 'total_pymnt_inv' 'total_rec_prncp']
total_rec_int ['total_rec_int']
total_rec_lat

  # Summary of Data Exploration

Raw dataset of Lending Club loans, spanning from  2007 to 2018 ("accepted_2007_to_2018Q4.csv") pulled from Kaggle.com
Dataset contains over 10 years of loan data, 2.26 million rows of data and 151 columns
There are 63 columns that have over 100,000 missing data points & 94 columns missing over 50,000 values

Loans that have been "Fully Paid" or "Charged Off" is an adequate size for ML models
 * Fully Paid loans: 1,076,751
 * Charged Off loans: 268,559
There appears to be an imbalance between the number of 'good' or "Fully Paid" loans and the 'bad' or "Charged Off" loans

Most data is in "integer" data type, there are 38 columns that are "string" or "date" data types, 
however many of these "string"/"date" columns are also missing large amounts of data ("na")

Much of the missing data will be 'dropped' using the dropna method, and columns/features that 
do not have any prediction power may be dropped after consulting the dataset's data dictionary 
to verify what data they contain and evaluate if the information can be known at the time of loan
application, as well as if the data has no predictive value

Running a correlation table between all of our data columns, we find that there is very high correlation 
between several columns/features which indicates the possibility of duplicate or redundant data